src/v/pandaproxy/schema_registry/service.cc

```diff
@@ -30,17 +31,40 @@
 #include <seastar/core/memory.hh>
 #include <seastar/core/std-coroutine.hh>
 #include <seastar/http/api_docs.hh>
+#include <seastar/http/exception.hh>
 
 namespace pandaproxy::schema_registry {
 
 using server = ctx_server<service>;
 
 template<typename Handler>
 auto wrap(ss::gate& g, one_shot& os, Handler h) {
-    return [&g, &os, _h{std::move(h)}](
+    return [&g, &os, h{std::move(h)}](
              server::request_t rq,
              server::reply_t rp) -> ss::future<server::reply_t> {
-        auto h{_h};
+        rq.authn_method = rq.service().config().authn_method(
+          rq.req->get_listener_idx());
+
+        if (rq.authn_method == config::rest_authn_method::http_basic) {
+            // Will throw 400 & 401 if auth fails
+            auto auth_result = rq.service().authenticator().authenticate(
+              *rq.req);
+            try {
+                auth_result.require_authenticated();
+            } catch (const ss::httpd::base_exception& ex) {
+                // The failure from require_authenticated() will throw 401
+                // instead of 403. So convert the exception to 403.
+                if (
+                  ex.status() == ss::httpd::reply::status_type::unauthorized) {
+                    throw ss::httpd::base_exception(
+                      ex.what(), ss::httpd::reply::status_type::forbidden);
+                } else {
+                    // Something else went wrong
+                    throw;
+                }
+            }
+        }
```

Comment:
This looks familiar, can the code be shared?

Something like (this might be a bit out of date): https://github.com/redpanda-data/redpanda/commit/e1f92a2025392e0f4b4d7b8a717fbd7326964335 and https://github.com/redpanda-data/redpanda/commit/e73d5b7233d8f7da54fb36cf32feb3bc29d43c75
And then it's possible to add some nice helpers: https://github.com/redpanda-data/redpanda/commit/7d82be4e2d38607c7e6a4f7f57a53db24b877f51
Which can simplify the handlers: https://github.com/redpanda-data/redpanda/commit/eccfeed81ac741bf68b53b4fb35f0787e3eebdfe

But I'll leave it to you to figure out the best way to reduce the dupolication.

Reply:
I like your context server changes but I think I'll put them in #6626 after we get the initial feature PRs out